security/pkg/nodeagent/caclient/providers/citadel/client.go

```diff
@@ -16,77 +16,107 @@ package caclient
-	// May also be used in K8S to use a Secret to bootstrap (as a 'refresh key'), but use short-lived tokens
-	// with extra SAN (labels, etc) in data path.
-	// TODO: move to main, stop using directly - security.Options instead !
-	ProvCert = env.RegisterStringVar("PROV_CERT", "",
-		"Set to a directory containing provisioned certs, for VMs").Get()
 )
 
 type citadelClient struct {
-	caEndpoint    string
-	enableTLS     bool
-	caTLSRootCert []byte
-	client        pb.IstioCertificateServiceClient
-	clusterID     string
-	conn          *grpc.ClientConn
+	caEndpoint string
+	client     *signingapi.Client
+	secOpts    *security.Options
 }
 
 // NewCitadelClient create a CA client for Citadel.
-func NewCitadelClient(endpoint string, tls bool, rootCert []byte, clusterID string) (security.Client, error) {
+func NewCitadelClient(endpoint string, tls bool, rootCert string, secOpts *security.Options) (security.Client, error) {
+	clientTLSSettings := &v1alpha3.ClientTLSSettings{
+		CaCertificates: rootCert,
+	}
+
+	if tls {
+		clientTLSSettings.Mode = v1alpha3.ClientTLSSettings_SIMPLE
+	} else {
+		clientTLSSettings.Mode = v1alpha3.ClientTLSSettings_DISABLE
+	}
+
+	// Initial implementation of citadel hardcoded the SAN to 'istio-citadel'. For backward compat, keep it.
+	// TODO: remove this once istiod replaces citadel.
+	// External CAs will use their normal server names.
+	if strings.Contains(endpoint, "citadel") {
+		clientTLSSettings.Sni = caServerName
+	}
+	// For debugging on localhost (with port forward)
+	// TODO: remove once istiod is stable and we have a way to validate JWTs locally
+	if strings.Contains(endpoint, "localhost") {
+		clientTLSSettings.Sni = "istiod.istio-system.svc"
+	}
+
+	// Connect with pre provisioning TLS Client Certificates
+	if secOpts.ProvCert != "" {
+		clientTLSSettings.PrivateKey = path.Join(secOpts.ProvCert, "key.pem")
+		clientTLSSettings.ClientCertificate = path.Join(secOpts.ProvCert, "cert-chain.pem")
+		clientTLSSettings.Mode = v1alpha3.ClientTLSSettings_MUTUAL
+	}
+
+	signingAPIClient := signingapi.New(endpoint, clientTLSSettings, requestTimeout)
+
 	c := &citadelClient{
-		caEndpoint:    endpoint,
-		enableTLS:     tls,
-		caTLSRootCert: rootCert,
-		clusterID:     clusterID,
+		caEndpoint: endpoint,
+		client:     signingAPIClient,
+		secOpts:    secOpts,
 	}
 
-	conn, err := c.buildConnection()
+	err := c.client.Connect(nil)
 	if err != nil {
 		citadelClientLog.Errorf("Failed to connect to endpoint %s: %v", endpoint, err)
 		return nil, fmt.Errorf("failed to connect to endpoint %s", endpoint)
 	}
-	c.conn = conn
-	c.client = pb.NewIstioCertificateServiceClient(conn)
 	return c, nil
 }
 
 // CSR Sign calls Citadel to sign a CSR.
 func (c *citadelClient) CSRSign(ctx context.Context, reqID string, csrPEM []byte, token string,
 	certValidTTLInSec int64) ([]string /*PEM-encoded certificate chain*/, error) {
+	csrMetadata := &signingapi.Metadata{
```

Comment:
We need some contract on what metadata is sent. That could just be a comment I suppose

Reply:
Otherwise someone will come and do some "code cleanup"

Reply:
Thanks @howardjohn